substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -1052,17 +1125,34 @@ impl<T: Config> Pallet<T> {
 	/// `Validators` or `VoterList` outside of this function is almost certainly
 	/// wrong.
 	pub fn do_remove_validator(who: &T::AccountId) -> bool {
-		let outcome = if Validators::<T>::contains_key(who) {
-			Validators::<T>::remove(who);
-			let _ = T::VoterList::on_remove(who).defensive();
-			true
-		} else {
-			false
+		let outcome = match (Validators::<T>::contains_key(who), T::TargetList::contains(who)) {
```

Comment:
Do these event push needs to be so complicated? Staking pallet should not ideally depend on/know anything about how TargetList and stake tracker works. If it removes validator, it should emit an event. 
I am guessing there might be some edge cases that you are trying to cover? We should try at least to simplify these.

Reply:
The reason for the staking pallet checking the state of the target list is to ensure that the validator is idle before removing it. Being idle (or chilling) means that we should first remove the self-stake approvals from the target list, and only then remove the target (if possible). This state can only be checked at the stke-tracker leve, unfortunately.

We can do one of two things here:
 1. I this check as a `debug_assert!()` instead and remove the call to `T::TargetList` from the main logic. This assumes that the staking pallet will always call chill before remove.
 2. make the call to `Self::do_chill_validator` idempotent and call it everytime we remove a validator.

I will try to implement 2. so that we don't need to rely on the data of the stake-tracker on the staking side.